lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp

```diff
@@ -68,6 +65,50 @@ bool TTNNLayoutAttr::hasInterleavedDRAMTensorMemoryLayout() const {
          (getMemLayout().getValue() == TensorMemoryLayout::Interleaved);
 }
 
+// Calculate the logical shape of the shard.
+// Shard is defined as a piece of the tensor that is mapped to a single grid
+// core. It is assumed that the TensorMemoryLayout is sharded.
+llvm::SmallVector<int64_t>
+TTNNLayoutAttr::calculateLogicalShardShapeForSharding(
+    ArrayRef<int64_t> tensorShape, mlir::AffineMap linear, GridAttr grid) {
+  assert(linear.getNumResults() == grid.getShape().size());
+  mlir::SmallVector<std::int64_t> logicalShape =
+      ttmlir::utils::evalShape(linear, tensorShape);
+  mlir::SmallVector<std::int64_t> shardShape(linear.getNumResults());
+  for (unsigned i = 0; i < linear.getNumResults(); ++i) {
+    shardShape[i] =
+        (logicalShape[i] + grid.getShape()[i] - 1) / grid.getShape()[i];
+  }
+  return shardShape;
+}
+
+// Calculate the logical shape of the shard.
+// Shard is defined as a piece of the tensor that is mapped to a single grid
+// core. It is assumed that the TensorMemoryLayout is interleaved.
+llvm::SmallVector<int64_t>
+TTNNLayoutAttr::calculateLogicalShardShapeForInterleaved(
+    ArrayRef<int64_t> tensorShape, mlir::Type elementType,
+    mlir::AffineMap linear, mlir::tt::GridAttr grid) {
+  assert(linear.getNumResults() == grid.getShape().size());
+  assert(mlir::isa<mlir::tt::TileType>(elementType));
+
+  mlir::SmallVector<std::int64_t> logicalShape =
+      ttmlir::utils::evalShape(linear, tensorShape);
+  mlir::SmallVector<std::int64_t> logicalTiledShape =
+      mlir::cast<mlir::tt::TileType>(elementType).getTiledShape(logicalShape);
+  uint64_t numOfTiles =
+      std::accumulate(logicalTiledShape.begin(), logicalTiledShape.end(), 1,
+                      std::multiplies<std::int64_t>());
+  uint64_t numOfGridUnits =
+      std::accumulate(grid.getShape().begin(), grid.getShape().end(), 1,
```

Comment:
Maybe leave an assert here that grid is equal to device max grid? Or at least a comment. I'm not sure that l1 interleaved is supported for non max grid configurations.

Reply:
We have to check this, but I suspect that l1 is supported for non max grid configuration.